packages/fxa-auth-server/fxa-oauth-server/lib/client.js

```diff
@@ -0,0 +1,107 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const crypto = require('crypto');
+const buf = require('buf').hex;
+const Joi = require('joi');
+
+const AppError = require('./error');
+const validators = require('./validators');
+const db = require('./db');
+const encrypt = require('./encrypt');
+const logger = require('./logging')('client');
+
+// Client credentials can be provided in either the Authorizxation header
+// or the request body, but not both. These are some re-useable validators
+// to assert that requirement.
+module.exports.clientAuthValidators = {
+  headers: Joi.object({
+    authorization: Joi.string()
+      .regex(validators.BASIC_AUTH_HEADER)
+      .optional(),
+  }).options({ allowUnknown: true }),
+
+  clientId: validators.clientId.when('$headers.authorization', {
+    is: Joi.string().required(),
+    then: Joi.forbidden(),
+  }),
+
+  clientSecret: validators.clientSecret.when('$headers.authorization', {
+    is: Joi.string().required(),
+    then: Joi.forbidden(),
+  }),
+};
+
+// Authenticate a request made by an OAuth client, using credentials from
+// either the Authorization header or request body parameters.
+//
+// If successful, returns an object with info about the client.
+// If unsuccessful, throws an error.
+//
+module.exports.authenticateClient = async function authenticateClient(
```

Comment:
Previously we only needed this logic in `/token`, but it's now moved to a shared helper.